aws/resource_aws_network_interface_sg_attachment.go

```diff
@@ -0,0 +1,181 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"reflect"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsNetworkInterfaceSGAttachment() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsNetworkInterfaceSGAttachmentCreate,
+		Read:   resourceAwsNetworkInterfaceSGAttachmentRead,
+		Delete: resourceAwsNetworkInterfaceSGAttachmentDelete,
+		Schema: map[string]*schema.Schema{
+			"security_group_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"network_interface_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsNetworkInterfaceSGAttachmentCreate(d *schema.ResourceData, meta interface{}) error {
+	mk := "network_interface_sg_attachment_" + d.Get("network_interface_id").(string)
+	awsMutexKV.Lock(mk)
+	defer awsMutexKV.Unlock(mk)
```

Comment:
It would be great if the mutexes were inside `attachSecurityGroupToInterface` & `detachSecurityGroupFromInterface` so we could encapsulate the logic of detaching/attaching SG and avoid forgetting mutexes when these functions get called elsewhere in the codebase.

That said I understand you probably put them here to keep it locked until we read the data back in `resourceAwsNetworkInterfaceSGAttachmentRead` below, which is reasonable due to eventual consistency issues, so I guess you can ignore me. 🤷‍♂️